File: 19 remove nth node from end of list.cpp
// https://leetcode.com/problems/remove-nth-node-from-end-of-list
// TS:O(n) SC:O(1)
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode * node = head;
        ListNode * toDeleteNode = head;
        ListNode * previousNode = head;
        while(node != NULL){
            if (n <= 0){ //it is assumed we have n <= size
                previousNode = toDeleteNode;
                toDeleteNode = toDeleteNode->next;
            }
            n--;
            node = node->next;
        }
        if (head == toDeleteNode){
            head = toDeleteNode->next;
        }
        else{
            previousNode->next = toDeleteNode->next;   
        }
        delete toDeleteNode;
        
        return head;
    }
};

File: 19 remove nth node from end of list.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list
# TC: O(n) SC: O(1)
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        front = head
        while front is not None and n > 0:
            front = front.next
            n -= 1
        dummy = ListNode(0, head)
        prev = dummy
        while front is not None:
            front = front.next
            prev = prev.next
        prev.next = prev.next.next
        return dummy.next